bot/cogs/help.py

```diff
@@ -68,499 +67,307 @@ class HelpSession:
         cog_matches = []
         description = None
-        for cog in self._bot.cogs.values():
-            if hasattr(cog, "category") and cog.category == query:
+        for cog in bot.cogs.values():
+            if hasattr(cog, "category") and cog.category == command:
                 cog_matches.append(cog)
                 if hasattr(cog, "category_description"):
                     description = cog.category_description
 
-        # Try to search by cog name if no categories match.
-        if not cog_matches:
-            cog = self._bot.cogs.get(query)
+        if cog_matches:
+            category = Category(name=command, description=description, cogs=cog_matches)
+            await self.send_category_help(category)
+            return
 
-            # Don't consider it a match if the cog has a category.
-            if cog and not hasattr(cog, "category"):
-                cog_matches = [cog]
+        # it's either a cog, group, command or subcommand, let super deal with it
+        await super().command_callback(ctx, command=command)
 
-        if cog_matches:
-            cog = cog_matches[0]
-            cmds = (cog.get_commands() for cog in cog_matches)  # Commands of all cogs
+    def get_all_help_choices(self) -> set:
+        """
+        Get all the possible options for getting help in the bot.
 
-            return Cog(
-                name=cog.category if hasattr(cog, "category") else cog.qualified_name,
-                description=description or cog.description,
-                commands=tuple(itertools.chain.from_iterable(cmds))  # Flatten the list
-            )
+        These include:
+        - Category names
+        - Cog names
+        - Group command names (and aliases)
+        - Command names (and aliases)
+        - Subcommand names (with parent group and aliases for subcommand, but not including aliases for group)
+
+        Options and choices are case sensitive.
+        """
+        # first get all commands including subcommands and full command name aliases
+        choices = set()
+        for c in self.context.bot.walk_commands():
+            # the the command or group name
+            choices.add(str(c))
+
+            # all aliases if it's just a command
+            choices.update(n for n in c.aliases if isinstance(c, Command))
 
-        self._handle_not_found(query)
+            # else aliases with parent if group. we need to strip() in case it's a Command and `full_parent` is None,
+            # otherwise we get 2 commands: ` help` and normal `help`.
+            # We could do case-by-case with f-string but this is the cleanest solution
+            choices.update(f"{c.full_parent_name} {a}".strip() for a in c.aliases)
 
-    def _handle_not_found(self, query: str) -> None:
+            # all cog names
+        choices.update(self.context.bot.cogs)
+
+        # all category names
+        choices.update(getattr(n, "category", None) for n in self.context.bot.cogs if hasattr(n, "category"))
```

Comment:
Having `getattr(n, "category", None) ... if hasattr(n, "category")` seems redundant to me.

```suggestion
        choices.update(n.category for n in self.context.bot.cogs if hasattr(n, "category"))
```